File: src/main/java/toker/panel/entity/CraftingRecipeItemRequirement.java
package toker.panel.entity;

import com.fasterxml.jackson.annotation.JsonIgnore;
import toker.panel.entity.pk.CraftingRecipeItemRequirementPK;

import javax.persistence.*;
import java.util.Objects;

@Entity
@Table(name = "crafting_recipe_item_requirement")
@IdClass(CraftingRecipeItemRequirementPK.class)
public class CraftingRecipeItemRequirement {

    private CraftingRecipe craftingRecipe;
    private Item item;
    private Integer amount;

    @Id
    @ManyToOne
    @JoinColumn(name = "recipe_id", referencedColumnName = "id", nullable = false)
    @JsonIgnore
    public CraftingRecipe getCraftingRecipe() {
        return craftingRecipe;
    }

    public void setCraftingRecipe(CraftingRecipe craftingRecipeByRecipeId) {
        this.craftingRecipe = craftingRecipeByRecipeId;
    }

    @Id
    @ManyToOne
    @JoinColumn(name = "item_id", referencedColumnName = "id", nullable = false)
    public Item getItem() {
        return item;
    }

    public void setItem(Item itemByItemId) {
        this.item = itemByItemId;
    }

    @Column(name = "amount", nullable = false)
    public Integer getAmount() {
        return amount;
    }

    public void setAmount(Integer amount) {
        this.amount = amount;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        CraftingRecipeItemRequirement that = (CraftingRecipeItemRequirement) o;
        return Objects.equals(craftingRecipe, that.craftingRecipe) &&
                Objects.equals(item, that.item);
    }

    @Override
    public int hashCode() {
        return Objects.hash(craftingRecipe, item);
    }
}

File: src/main/java/toker/panel/entity/CraftingRecipeItemRequirement.kt
package toker.panel.entity

import com.fasterxml.jackson.annotation.JsonIgnore
import toker.panel.entity.pk.CraftingRecipeItemRequirementPK
import javax.persistence.*

@Entity
@Table(name = "crafting_recipe_item_requirement")
@IdClass(CraftingRecipeItemRequirementPK::class)
class CraftingRecipeItemRequirement(
        @Id
        @ManyToOne
        @JoinColumn(name = "recipe_id", referencedColumnName = "id", nullable = false)
        @JsonIgnore
        var craftingRecipe: CraftingRecipe? = null,

        @Id
        @ManyToOne
        @JoinColumn(name = "item_id", referencedColumnName = "id", nullable = false)
        var item: Item? = null,

        @Column(name = "amount", nullable = false)
        var amount: Int? = null
) {
        override fun equals(other: Any?): Boolean {
                if (this === other) return true
                if (javaClass != other?.javaClass) return false
                other as CraftingRecipeItemRequirement
                if (craftingRecipe != other.craftingRecipe) return false
                if (item != other.item) return false
                return true
        }

        override fun hashCode(): Int {
                var result = craftingRecipe?.hashCode() ?: 0
                result = 31 * result + (item?.hashCode() ?: 0)
                return result
        }
}